### PR TITLE
[java-gen] Adding support for default values

### PR DESCRIPTION
**Pull request by fabiobrz**
Draft/attempt to support default values generation. 

Fixes https://github.com/fabric8io/kubernetes-client/issues/3869

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JDefaultValue.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    this.key = key;
+    this.type = type;
+    this.defaultValueNode = defaultValueNode;
+  }
+
+  public String getKey() {
+    return key;
+  }
+
+  public static JDefaultValue fromJsonSchema(String key, String type, AbstractJSONSchema2Pojo source,
```
Comment (andreaTP) [inline java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JDefaultValue.java:38]: As discussed, we should try to leverage the tree generated already by `AbstractJSONSchema2Pojo`.

Possibly, the default value can be added as an additional field in `GeneratorResult` and expanded when needed by `JObject` during `generateJava` (I'm open to other alternatives though! 🙂  )

**Review by andreaTP — COMMENTED** [review]
Thanks a lot for your interest @fabiobrz !

Looks like a great starting point!!!

**Review by andreaTP — CHANGES_REQUESTED** [review]
Thanks a lot @fabiobrz !
I added a few notes, will try to take a deeper look some time soon.

TL;DR:
 - the implementation for `nullable` can be separated (so that is faster to merge)
 - the implementation of `default` might require a few more iterations 🙂

**Comment by fabiobrz** [comment]
> Thanks a lot @fabiobrz ! I added a few notes, will try to take a deeper look some time soon.
> 
> TL;DR:
> 
>     * the implementation for `nullable` can be separated (so that is faster to merge)
> 
>     * the implementation of `default` might require a few more iterations slightly_smiling_face

Thanks for the review @andreaTP! :) 
I'll go through each of your comments to fix the code and yes, I'll extract the `Nullable` part to a separate PR. I'll keep you posted!

File: java-generator/it/src/it/cert-manager/expected/Auth.expected [hunk 2]
```diff
@@ -57,6 +59,7 @@ public class Auth implements io.fabric8.kubernetes.api.model.KubernetesResource
      */
     @com.fasterxml.jackson.annotation.JsonProperty("tokenSecretRef")
     @com.fasterxml.jackson.annotation.JsonPropertyDescription("TokenSecretRef authenticates with Vault by presenting a token.")
+    @com.fasterxml.jackson.annotation.JsonSetter(nulls=com.fasterxml.jackson.annotation.Nulls.SKIP)
```
Comment (andreaTP) [inline java-generator/it/src/it/cert-manager/expected/Auth.expected:62]: Before I forget, @metacosm and @manusa would you like to take a look at this encoding from the Jackson point of view and if we can rely on it in the crd-generator (eventually)?

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JDefaultValue.java (hunk 1 above)
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JDefaultValue.java]: In commits that are following this comment I removed the additional hierarchy that I created to represent instances of defaults, so I'll resolve this comment and move on with the latest review.

[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JDefaultValue.java]: As stated above, now the default generation logic is provided by the `AbstractJSONSchema2Pojo` hierarchy. Resolving this thread.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java [hunk 3]
```diff
@@ -39,11 +42,14 @@
 
   protected final String description;
   protected final Config config;
+  protected final boolean isNullable;
 
   public abstract String getType();
 
   public abstract GeneratorResult generateJava();
 
+  public abstract Expression generateDefaultValue(BlockStmt scope, final String propertyKey, JsonNode defaultValue);
```
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java:51]: Hi @andreaTP - Interesting topic here, I'd like to discuss.
IIUC, `AbstractJSONSchema2Pojo` is a _definition_, not an _instance_. 
As suche, there cannot be a representation for each and every default that you could have. A use case is represented by a CRD `array` of `string` instances, which default value would be `["a", "b", "c"]`.
The runtime tree would have a `JArray`, with _one `JPrimitive` as `nested`, hence you would not have any representation instance where to store the default value _occurrence_ at runtime.

This is the reson for passing `defaultValue` to the `AbstractJSONSchema2Pojo` instance that `JObject` is curently generated, i.e. `AbstractJSONSchema2Pojo` concrete instance _know_ how to render a default of their own type (hence the logic is fully encapsulated by the hierarchy levels) but they need the value to apply the process to.

WDYT? Once we get things clear about this, then we'll move to the other topics that you touched, i.e. the `parent` object role in the generation, and keys.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java [hunk 4]
```diff
@@ -95,16 +106,31 @@ public JObject(
...
           String nextSuffix = (config.getSuffixStrategy() == Config.Suffix.ALWAYS) ? classSuffix : "";
-
+          AbstractJSONSchema2Pojo fieldProp = AbstractJSONSchema2Pojo.fromJsonSchema(
+              field.getKey(),
+              field.getValue(),
+              nextPackagePath,
+              nextPrefix,
+              nextSuffix,
+              config);
+          // Let's have some additional criteria when generating default, i.e.: nodes with variable
```
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java:116]: Yes, I decided to use the following:

`x-kubernetes-preserve-unknown-fields: true`

as an additional discriminant to include the default value into the representation. My intent was to skip objets that don't have a strict definition for default generation. Take the following case, an CRD `object` with no props and `x-kubernetes-preserve-unknown-fields: true`, see `readinessProbe` in akka-microservices CRD. It has a default to set `periodSeconds` but the generated object won't have that field and the code which would be generated to initialize the default would reference a non-existing field. And it won't compile.

I must admit aìthat a better criteria would rely on previous validation, but we don't have it. Maybe I should adjust the proposed changes in order to use a different condition, i.e. when `properties` is not defined) to better identify the case where no structure definition is provided. WDYT?

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java [hunk 5]
```diff
@@ -208,6 +234,20 @@ public GeneratorResult generateJava() {
                   new StringLiteralExpr(
                       prop.getDescription().replace("\"", "\\\""))));
         }
+
+        if (!prop.isNullable) {
```
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java:238]: Done in https://github.com/fabric8io/kubernetes-client/pull/3977 and removed from this PR branch (will not be in the next push) - resolivng this thread.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JPrimitive.java [hunk 6]
```diff
@@ -38,4 +47,30 @@ public String getType() {
   public GeneratorResult generateJava() {
     return empty;
   }
+
+  @Override
+  public Expression generateDefaultValue(BlockStmt scope, final String propertyKey, JsonNode defaultValue) {
+    Expression expr;
+    if (!JPrimitiveNameAndType.STRING.getName().equals(this.type)) {
```
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JPrimitive.java:54]: +1, I'll check and see whether I can remove it, otherwise I'll convert it to a `switch`.

File: java-generator/core/src/test/resources/io/fabric8/java/generator/approvals/ApprovalTest.testAkkaMicroservicesCrd.approved.txt [hunk 7]
```diff
@@ -0,0 +1,668 @@
...
+    public org.test.v1.akkamicroservicespec.Resources getResources() {
+        return resources;
+    }
+
+    public void setResources(org.test.v1.akkamicroservicespec.Resources resources) {
+        this.resources = resources;
+    }
+
+    private org.test.v1.akkamicroservicespec.Resources init_resources() {
+        org.test.v1.akkamicroservicespec.Resources resources = new org.test.v1.akkamicroservicespec.Resources();
```
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/test/resources/io/fabric8/java/generator/approvals/ApprovalTest.testAkkaMicroservicesCrd.approved.txt:482]: > As an hint (possibly) we should be able to generate default values without declaring "intermediate" variables.
Worst case scenario we can declare only one, e.g. tmp , it should, maybe, look like:

Is there a reason for this? Using variables as proposed in this PR significantly reduces the logic needed to generate complex defaults, like objects one, have a look at the following experimental case I am using locally:
```
private com.lightbend.akka.v1.akkamicroservicespec.Resources init_resources() {
        com.lightbend.akka.v1.akkamicroservicespec.Resources resources = new com.lightbend.akka.v1.akkamicroservicespec.Resources();
        com.lightbend.akka.v1.akkamicroservicespec.resources.Limits limits = new com.lightbend.akka.v1.akkamicroservicespec.resources.Limits();
        limits.setMemory("1024Mi");
        limits.setCpu("");
        com.lightbend.akka.v1.akkamicroservicespec.resources.limits.Cat cat = new com.lightbend.akka.v1.akkamicroservicespec.resources.limits.Cat();
        cat.setFur("brown");
        com.lightbend.akka.v1.akkamicroservicespec.resources.limits.cat.Gender gender = new com.lightbend.akka.v1.akkamicroservicespec.resources.limits.cat.Gender();
        gender.setName("female");
        gender.setIdentifier("F");
        cat.setGender(gender);
        cat.setCode(123L);
        cat.setEuropean(true);
        limits.setCat(cat);
        resources.setLimits(limits);
        com.lightbend.akka.v1.akkamicroservicespec.resources.Requests requests = new com.lightbend.akka.v1.akkamicroservicespec.resources.Requests();
        requests.setMemory("1024Mi");
        requests.setCpu("1");
        resources.setRequests(requests);
        return resources;
    }
```
I really prefer this against a chain of `.get...()` calls, but maybe I didn't get your point?
About generating via `sundrio` I feel that adding another level wouldn't be needed at this point, but maybe you've got some justification for it that I don't see now?

Regarding: 
> have you checked that it respects the correct semantics?

I am not sure I get your point, here's the CRD fragment:

```
cpu:
                          description: 'While not recommended, this allows to set a CPU limit. See https://doc.akka.io/docs/akka/current/additional/deploying.html#resource-limits'
                          type: string
                          default: ""
```
 can you elaborate on what semantics are you relating to?

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java:116]: Definitely this should be done differently, anyway. I am going to fix the criteria evaluation by removing all those RTTI based conditions and casts.

File: java-generator/core/src/test/resources/io/fabric8/java/generator/approvals/ApprovalTest.testAkkaMicroservicesCrd.approved.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/test/resources/io/fabric8/java/generator/approvals/ApprovalTest.testAkkaMicroservicesCrd.approved.txt]: FIxed by passing the scope through the recursion.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java]: Fixed, now as far as `object`s are concerned, the default is generated only for those with a defined set of `properties`, otherwise the generated code for the default to be set would not compile against the generated class code.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java]: As msaid, this ismplementation is based on the existing graph, which is made of nodes that represent a schema, ot instances. _Default values_ are instances, and as such each class can provide the logic to code a default value generation, but the default value itself _must_ be passed in as an argument. The principal use case here is `JArray` or `JMap` that have one `nested` node to represent the element type they contain, but could have a default value which stores _N_ elements.

File: java-generator/it/src/it/cert-manager/expected/Auth.expected (hunk 2 above)
Comment (fabiobrz) [reply java-generator/it/src/it/cert-manager/expected/Auth.expected]: `Nulls` management implementation has been merged already separately, resolving.

**Comment by fabiobrz** [comment]
Hi @andreaTP - I revisited the implementation by following your suggestions and trying to address your concerns as much as possible.

I also provided a different implementation which is based on `Serialization::unmarshal`, as agreed.
That being said I think that this implementation has some advantages:
* doesn't depend on the kubernetes-client classes at runtime
* tests are more effective, considering that the generated code doesn't on deserialization at runtime
* addresses one issue with `object`s which are missing a defined set of `properties`, by skipping such onjects since the generated code for initializing default values would not compile against a generated class that doesn't have correspondent fields.

Feel free to let me know what you think.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JPrimitive.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JPrimitive.java:66]: I quickly looked and it seems to me that:

* `this.type` is a `String`, whould be good to switch on, but `JPrimitiveNameAndType::getName` isn't a constant value, hence cannot be used as a `case` label.
* `JPrimitiveNameAndType` constants (e.g.: `JPrimitiveNameAndType.STRING`) cannot be used to switch cases, hence.

But I might not see what you mean, feel free to suggest :slightly_smiling_face:

**Comment by andreaTP** [comment]
Hi @fabiobrz , thanks a lot for the effort! Really, really appreciated ❤️  🙂 .

I want to share my point of view regarding your points:

> doesn't depend on the kubernetes-client classes at runtime

We already depend on `kubernetes-client` for a number of very "structural" things, `CustomResource`, `Version` and `Group` annotations, `IntOrString`, just to mention a few.

> tests are more effective, considering that the generated code doesn't rely on runtime deserialization

I do believe that those tests are already covered in the Jackson codebase, also using `Serialization` we use the same mechanism to serialize/deserialize everything and we are not subject to edge-cases or differences for the end-user behaviour.

> addresses one issue with objects which are missing a defined set of properties, by skipping such onjects since the generated code for initializing default values would not compile against a generated class that doesn't have correspondent fields.

Here you are touching a point I forgot to verify in the other PR, can you please detail the invariants?
On top of my head I can think about only case that is when `x-kubernetes-preserve-unknown-fields` is defined, and we should have proper testing for this situation, but, using `Serialization` and plain Jackson this should be already correctly handled in the other PR (e.g. applying `JsonAnyGetter`/`JsonAnySetter`).

As you noticed I'm already biased toward the other implementation 🙂

**Comment by fabiobrz** [comment]
> As you noticed I'm already biased toward the other implementation slightly_smiling_face

Ok, thanks @andreaTP :slightly_smiling_face:  - let's wait what happens to https://github.com/fabric8io/kubernetes-client/pull/4019 and then we might close this one.

**Comment by fabiobrz** [comment]
Closing, since https://github.com/fabric8io/kubernetes-client/pull/4019 has been merged.
